include/deal.II/lac/sparse_matrix.h

```diff
@@ -1868,8 +1874,21 @@ SparseMatrix<number>::operator /= (const number factor)
 
 template <typename number>
 inline
-number SparseMatrix<number>::operator () (const size_type i,
-                                          const size_type j) const
+number &SparseMatrix<number>::operator () (const size_type i,
+                                           const size_type j) const
+{
+  Assert (cols != nullptr, ExcNotInitialized());
+  Assert (cols->operator()(i,j) != SparsityPattern::invalid_entry,
+          ExcInvalidIndex(i,j));
+  return val[cols->operator()(i,j)];
```

Comment:
I don't understand how this compiles -- you're in a `const` member function, so `val` is `const` but you're returning a non-`const` reference.

Reply:
hm, indeed, should have failed but it does not

Reply:
Do you have a test for this case?

Reply:
> Do you have a test for this case?

should be used in the test in this PR

Reply:
Then I can only assume that you are returning a reference to a temporary here. Either way, you probably want to change the return type if you haven't already.

Reply:
> if you haven't already.

i already did

Reply:
Here you could also just use the non-const version.

Reply:
@masterleinad 

not sure how would you do this. I get

> warning: all paths through this function will call itself [-Winfinite-recursion]

Reply:
@masterleinad there those two options: 

[1](https://stackoverflow.com/questions/123758/how-do-i-remove-code-duplication-between-similar-const-and-non-const-member-func) and [2](https://stackoverflow.com/questions/4047161/call-a-non-const-member-function-from-a-const-member-function/4047180#4047180)

is this what you had in mind?

Reply:
Yes, more or less. But I don't feel strongly about it (especially if we can't avoid a `const_cast`).

Reply:
No, let's not use a `const_cast`. This is the correct way of implementing the function.